### ISSUE TITLE
Avoid merge conflicts in pinning files

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#8958]
Alex flagged how it is common for there to be merge conflicts in `tools/{oldest_constraints,requirements}.txt`. I hope the frequency of these merge conflicts slows down now that the pinning system is largely finalized, however, I was thinking more about if there's something we can do to reduce their frequency.

I'd need to play with it more, but one idea I had was to keep Poetry's lock file around (it's named `poetry.lock` and we currently delete it) and then do something with `poetry lock --no-update`? I think this might work if we're wanting to make changes based on changes to `pyproject.toml`, but I don't think this would work if we just want to update a subset of the dependencies. Maybe we could artificially introduce constraints by requiring a higher version number, but this seems kind of awkward to me.

@adferrand, I'd be interested to hear if your ideas about this problem if you have any and manage to find the spare cycles to share them.

**Comment by ohemorange** [comment]
Can you explain this idea a little more? What would keeping the lock file around do, what would the process be for taking control of the lock to make updates?

**Comment by adferrand** [comment]
Yes I would also some elaboration on how we could do this. So far I imagine that we could generate dynamically the requirements files from poetry using the lock files that would be checked out in the repository.

I would be also interested to see some concrete examples where these conflicts happen, to understand more why.

That said, I do not think the approach to resolve these conflicts should be in the direction of finding a better format for the constraints that allows git to find automatically a solution to merge two branches without conflicts. I suppose here that your interest for the lock file is in the structured data (a toml file) that gives more opportunities to git to solve itself the conflicts than with a plain txt file.

Indeed the outcome of the requirements file makes sense as a whole, as the result of a balance between all constraints described in the pyproject.toml file, on the contrary to a code file where two independent modifications on two different functions can nmbe merged together. Basically here you want your version of the requirements in your PR to unconditionnally overwrite the requirements in the master branch, because you have calculated the new file with all constraints (including your new one) taken into account.

So maybe the solution is in how a new functionality is divided into several PRs to be introduced in master. If a functionality requires a new dependency, or a new version of a dependency, maybe we should create à first PR with the sole purpose of upgrading the requirements, then another PR to introduced code modification.

In general it may be desirable to always have dedicated PRs to upgrade the requirements, and these PRs should exist during a short period of time, and avoid to have two different PRs existing in parallel in the queue.

**Comment by bmw** [comment]
> What would keeping the lock file around do? What would the process be for taking control of the lock to make updates?

Poetry's lock file is essentially its representation of our .txt pinning files. For example, [here is the poetry.lock file for dns-lexicon](https://github.com/AnalogJ/lexicon/blob/65b09e99190ff170ffc27bc589e1e589c92a9d74/poetry.lock). In our pinning scripts, we have [poetry export its lock file to a format pip can understand](https://github.com/certbot/certbot/blob/1129d850d3257deae1c5de5b35c35c873d1fca7c/tools/pinning/common/export-pinned-dependencies.sh#L44).

`poetry lock` which we run to generate the lockfile accepts a `--no-update` flag whose help text says "Do not update locked versions, only refresh lock file." This would need to be investigated more, but briefly playing with it, I believe this flag makes poetry update its lockfile so that it pins and satisfies the dependencies in the `pyproject.toml` file without updating any dependencies unless it was necessary to do so.

I think the benefit of this is it minimizes the number of dependencies that are updated reducing merge conflicts. We could and may still want to sometimes run `poetry lock` without `--no-update` to prevent our pinned dependencies from becoming ancient.

Like I said in my initial post, I'm also not sure the best way to use this if we just want to update some of our dependencies. The only way I currently see to do this would be to (temporarily?) add a constraint on a higher version in `pyproject.toml` which seems awkward.

> I would be also interested to see some concrete examples where these conflicts happen, to understand more why.

I think it was https://github.com/certbot/certbot/pull/8928 that caused this issue to be raised.

I created this issue because it was causing a bit of frustration and I had an idea that may help. With that said, I also would like to point out that [our old system updated all dependencies like this too](https://github.com/certbot/certbot/blob/v1.0.0/letsencrypt-auto-source/rebuild_dependencies.py) and I think merge conflicts here may slow down dramatically since we're no longer working on this system. If this is causing any significant frustration for people, I think we should definitely try and do something about it, but I personally think it may be a bit premature to tunnel down into this issue and add any more complexity to our system here.
